routing/payment_session.go

```diff
@@ -161,50 +168,87 @@ func (p *paymentSession) RequestRoute(maxAmt, feeLimit lnwire.MilliSatoshi,
 		PaymentAddr:       p.payment.PaymentAddr,
 	}
 
-	// We'll also obtain a set of bandwidthHints from the lower layer for
-	// each of our outbound channels. This will allow the path finding to
-	// skip any links that aren't active or just don't have enough bandwidth
-	// to carry the payment. New bandwidth hints are queried for every new
-	// path finding attempt, because concurrent payments may change
-	// balances.
-	bandwidthHints, err := p.getBandwidthHints()
-	if err != nil {
-		return nil, err
-	}
-
 	finalHtlcExpiry := int32(height) + int32(finalCltvDelta)
 
-	path, err := p.pathFinder(
-		&graphParams{
-			graph:           ss.Graph,
-			additionalEdges: p.additionalEdges,
-			bandwidthHints:  bandwidthHints,
-		},
-		restrictions, &ss.PathFindingConfig,
-		ss.SelfNode.PubKeyBytes, p.payment.Target,
-		maxAmt, finalHtlcExpiry,
-	)
-	if err != nil {
-		return nil, err
-	}
-
-	// With the next candidate path found, we'll attempt to turn this into
-	// a route by applying the time-lock and fee requirements.
-	sourceVertex := route.Vertex(ss.SelfNode.PubKeyBytes)
-	route, err := newRoute(
-		sourceVertex, path, height,
-		finalHopParams{
-			amt:         maxAmt,
-			cltvDelta:   finalCltvDelta,
-			records:     p.payment.DestCustomRecords,
-			paymentAddr: p.payment.PaymentAddr,
-		},
-	)
-	if err != nil {
-		// TODO(roasbeef): return which edge/vertex didn't work
-		// out
-		return nil, err
+	for {
+		// We'll also obtain a set of bandwidthHints from the lower
+		// layer for each of our outbound channels. This will allow the
+		// path finding to skip any links that aren't active or just
+		// don't have enough bandwidth to carry the payment. New
+		// bandwidth hints are queried for every new path finding
+		// attempt, because concurrent payments may change balances.
+		bandwidthHints, err := p.getBandwidthHints()
```

Comment:
Moving the comment here, to prevent a very long list of "top-level" comments in github.

>Should we create an issue to track this? The sync handoff is also assumed by the payment loop, so we should probably fix this at one point...

I don't know if it is really an issue or merely an optimization. Although the boundary between the two is vague. I did think a bit about solutions. Maybe it is possible to hand off the htlc to the link directly, cut out the switch and mailbox. Or does the switch serve an essential purpose for locally initiated payments?

Reply:
If it is something we don't want to fix, then the whole launch then collect logic in the payment loop can be greatly simplified I think.

I agree it is an optimization, since the pathfinding will just be retried, but would be nice to move the assumptions to either extreme: assume the bandwidth hints tell you nothing about in-flight shards (and we could pass in a list of active shards, that can also help pathfinding on non-local channels) or assume they tell you everything you need to know (the current assumption).

Reply:
> Maybe it is possible to hand off the htlc to the link directly, cut out the switch and mailbox. Or does the switch serve an essential purpose for locally initiated payments?

Yeah, didn't look into how to actually solve it, but I was think something like an ACK channel to wait on.

Reply:
Issue created: https://github.com/lightningnetwork/lnd/issues/4181